ibis/backends/base/sql/__init__.py

```diff
@@ -9,3 +18,200 @@ class BaseSQLBackend(BaseBackend):
     """
     Base backend class for backends that compile to SQL.
     """
+
+    compiler = Compiler
+    table_class = ops.DatabaseTable
+    table_expr_class = ir.TableExpr
+
+    def table(self, name, database=None):
+        """Create a table expression.
+
+        Create a table expression that references a particular table in the
+        database.
+
+        Parameters
+        ----------
+        name : string
+        database : string, optional
+
+        Returns
+        -------
+        table : TableExpr
+        """
+        qualified_name = self._fully_qualified_name(name, database)
+        schema = self.get_schema(qualified_name)
+        node = self.table_class(qualified_name, schema, self)
+        return self.table_expr_class(node)
+
+    def _fully_qualified_name(self, name, database):
+        # XXX
+        return name
+
+    def sql(self, query):
+        """Convert a SQL query to an Ibis table expression.
+
+        Parameters
+        ----------
+        query : string
+
+        Returns
+        -------
+        table : TableExpr
+        """
+        # Get the schema by adding a LIMIT 0 on to the end of the query. If
+        # there is already a limit in the query, we find and remove it
+        limited_query = f'SELECT * FROM ({query}) t0 LIMIT 0'
+        schema = self._get_schema_using_query(limited_query)
+        return ops.SQLQueryResult(query, schema, self).to_expr()
+
+    def raw_sql(self, query: str, results=False):
+        """Execute a given query string.
+
+        Could have unexpected results if the query modifies the behavior of
+        the session in a way unknown to Ibis; be careful.
+
+        Parameters
+        ----------
+        query : string
+          DML or DDL statement
+
+        Returns
+        -------
+        Backend cursor
+        """
+        # TODO results is unused, it can be removed
+        # (requires updating Impala tests)
+        # TODO `self.con` is assumed to be defined in subclasses, but there
+        # is nothing that enforces it. We should find a way to make sure
+        # `self.con` is always a DBAPI2 connection, or raise an error
```

Comment:
What is the precise procedure for determining whether an object is a DBAPI2 connection?